presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java

```diff
@@ -438,13 +439,36 @@ protected TupleDescriptor visitInsert(Insert insert, AnalysisContext context)
         accessControl.checkCanInsertIntoTable(session.getIdentity(), targetTable);
         analysis.setInsertTarget(targetTableHandle.get());
 
-        List<ColumnMetadata> columns = metadata.getTableMetadata(session, targetTableHandle.get()).getColumns();
-        Iterable<Type> tableTypes = columns.stream()
-                .filter(column -> !column.isHidden())
-                .map(ColumnMetadata::getType)
-                .collect(toImmutableList());
+        List<ColumnMetadata> allTableColumns = metadata.getTableMetadata(session, targetTableHandle.get()).getColumns();
+        final Iterable<Type> tableTypes;
+        if (insert.getColumns().isPresent()) {
+            List<String> insertColumns = insert.getColumns().get();
+            analysis.setInsertColumns(insert.getColumns().get());
 
-        Iterable<Type> queryTypes = transform(descriptor.getVisibleFields(), Field::getType);
+            if (new HashSet<>(insertColumns).size() != insertColumns.size()) {
+                throw new SemanticException(DUPLICATE_COLUMN_NAME, insert, "Insert columns have at least one duplicated column");
+            }
+
+            tableTypes = insertColumns.stream()
```

Comment:
This nested stream code is really hard to read. Let's do it as a normal for-each loop.